### DISCUSSION TITLE
add short-circuit evaluation

### DISCUSSION THREAD

**Discussion by theDapperRaven** [godotengine/godot-proposals#10162] [Scripting]
I usually code in python and one thing from python that I miss in GDScript (other than F strings) is its short-circuit evaluation. 

If you don't know what short-circuit evaluation is, here is a python example to show it. 
```py
value = None
variable = value or 'Other'
```
If `value` is truthy, then `variable` will be defined as `value`. If not, it will be the string "Other"
In this case, None (python's null) is not truthy, so it will be considered False. `variable` ends up as "Other"

What do you guys think? I think it would be a handy little feature. 
Here is how it would be used in gdscpt
```gdscript
@export var my_resource: EnemyResource
@export var hitpoint_overwrite: int = null

func _ready():
    var max_health = hitpoint_overwrite or my_resource.hitpoints
    # if no override is added, it uses its default as seen in its custom resource
```
Nice and neat, eh?
currently, we need to do:
```gdscript
@export var my_resource: EnemyResource
@export var hitpoint_overwrite: int = null

func _ready():
    var max_health = hitpoint_overwrite if hitpoint_overwrite else my_resource.hitpoints
    # alternatively:
    if hitpoint_overwrite:
        max_health = hitpoint_overwrite
    else:
        max_health = my_resource.hitpoints
```
I suppose that it is readable, but a bit cluttered. Also a good deal longer. 

One issue I saw is that currently in GDScript, the `or` keyword is seen as a comparison operator in definitions. If short-circuit evaluation is added, this functionality would need to be changed. 
```gdscript
var a = null
var b = a or 'default'
# this makes b true, since a string is truthy.
# this behaviour also occurs for "and" and "not"
```
I cant think of a use case for this. Care to enlighten me, folks?

In conclusion, I think short-circuit evaluation would make code much cleaner and more readable. I don't know how to implement this, nor do I know the engine well enough to try, so I made this a discussion. 
What are your thoughts?

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/1321.

> In this case, None (python's null) is not truthy, so it will be considered False. variable ends up as "Other"

For reference, this already works in GDScript. This is also called the [Elvis operator](https://en.wikipedia.org/wiki/Elvis_operator) in some languages. It's like the null coalescing operator, but it falls back to the specified value whenever the value is falsy.

**Reply by theDapperRaven** [reply]
Oh, I was unaware of this!

> For reference, this already works in GDScript. This is also called the Elvis operator in some languages.

Really? I could not get it to work, nor can I find reference to it in the docs or online.

**Reply by Mickeon** [reply]
Calinou is potentially referring to an inline `if` (`var b = a if a else "Default"`)

**Reply by Calinou** [reply]
> Calinou is potentially referring to an inline if (var b = a if a else "Default")

Yes, that's the long form of what the Elvis operator expands to.

I didn't realize `a or "Default"` won't work though. Instead, it evaluates to a boolean expression which is always `true`, since `"Default"` is a non-empty string. Changing this behavior would break compatibility with existing projects.

**Reply by theDapperRaven** [reply]
Unfortunately yes. So a keyword like "otherwise" would be useful.
